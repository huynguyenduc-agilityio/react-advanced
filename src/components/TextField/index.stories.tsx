import { Flex } from '@chakra-ui/react';
import { Meta, StoryObj } from '@storybook/react';

import TextField from '.';
import { FaUser } from 'react-icons/fa';

const defaultProps = {
  label: 'Full name',
  placeholder: 'Please enter name',
  isError: false,
  errorMessages: '',
  isRequired: false,
  icon: FaUser,
  onChange: () => {},
};

const meta = {
  title: 'Components/TextField',
  component: TextField,
  parameters: {
    layout: 'centered',
  },

  tags: ['autodocs'],

  render: (args) => (
    <Flex justifyContent="center" w="1000px" p={9} bgColor="spaceBlue">
      <TextField {...args} />
    </Flex>
  ),
} satisfies Meta<typeof TextField>;

export default meta;

type Story = StoryObj<typeof TextField>;

export const Default: Story = {
  args: { ...defaultProps },
};

export const WithError: Story = {
  args: {
    ...defaultProps,
    isError: true,
    errorMessages: 'This field is required',
  },
};

export const WithoutIcon: Story = {
  args: {
    ...defaultProps,
    icon: undefined,
  },
};

export const RequiredField: Story = {
  args: {
    ...defaultProps,
    isRequired: true,
  },
};
