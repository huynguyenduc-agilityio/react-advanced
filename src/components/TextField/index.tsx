import { ChangeEvent, memo } from 'react';
import { IconType } from 'react-icons';
import {
  Box,
  Flex,
  FormControl,
  FormErrorMessage,
  FormLabel,
  Icon,
  Input,
  InputProps,
} from '@chakra-ui/react';

type TTextFieldProps = Omit<InputProps, 'onChange'> & {
  label: string;
  icon?: IconType;
  isRequired?: boolean;
  isError?: boolean;
  errorMessages?: string;
  onChange: (value: string) => void;
};

const TextField = ({
  label,
  isRequired = false,
  isError = false,
  errorMessages = 'Default error',
  icon,
  onChange,
  ...rest
}: TTextFieldProps) => {
  const handleChangeValue = (e: ChangeEvent<HTMLInputElement>): void =>
    onChange(e.target.value);

  return (
    <FormControl my={6} isRequired={isRequired} isInvalid={isError}>
      <Flex align="start">
        <Flex align="center" mt={1}>
          {icon && <Icon as={icon} color="white" boxSize={3} />}
          <FormLabel
            fontSize="sm"
            color="white"
            w="168px"
            textAlign="left"
            mb={0}
            ml={1}
          >
            {label}
          </FormLabel>
        </Flex>
        <Box flex="1">
          <Input
            variant="primary"
            onChange={handleChangeValue}
            {...rest}
            isInvalid={isError}
          />
          {isError && (
            <FormErrorMessage color="red" fontSize="xs">
              {errorMessages}
            </FormErrorMessage>
          )}
        </Box>
      </Flex>
    </FormControl>
  );
};

export default memo(TextField);
