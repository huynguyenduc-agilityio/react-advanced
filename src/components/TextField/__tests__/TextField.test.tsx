import { render, screen, fireEvent } from '@testing-library/react';
import { FaUser } from 'react-icons/fa';

import TextField from '..';

describe('TextField Component', () => {
  const mockOnChange = jest.fn();

  const commonProps = {
    label: 'Username',
    onChange: mockOnChange,
    placeholder: 'Enter your username',
  };

  it('renders the component with label and input field', () => {
    render(<TextField {...commonProps} />);

    // Check if the label and input are rendered
    expect(screen.getByLabelText(/username/i)).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText(/enter your username/i),
    ).toBeInTheDocument();
  });

  it('renders with an icon if provided', () => {
    render(<TextField {...commonProps} icon={FaUser} />);

    const icon = document.querySelector('.chakra-icon');

    // Check if the icon is rendered
    expect(screen.getByLabelText(/username/i)).toBeInTheDocument();
    expect(icon).toHaveClass('chakra-icon');
  });

  it('calls onChange when input value changes', () => {
    render(<TextField {...commonProps} />);

    const input = screen.getByLabelText(/username/i);
    fireEvent.change(input, { target: { value: 'new value' } });

    // Verify onChange is called with the new value
    expect(mockOnChange).toHaveBeenCalledTimes(1);
    expect(mockOnChange).toHaveBeenCalledWith('new value');
  });

  it('shows error message when isError is true', () => {
    render(
      <TextField
        {...commonProps}
        isError={true}
        errorMessages="Invalid input"
      />,
    );

    // Check if the error message is displayed
    expect(screen.getByText(/invalid input/i)).toBeInTheDocument();
    expect(screen.getByText(/invalid input/i)).toHaveStyle({ color: 'red' });
  });

  it('does not show error message when isError is false', () => {
    render(
      <TextField
        {...commonProps}
        isError={false}
        errorMessages="Invalid input"
      />,
    );

    // Ensure no error message is displayed
    expect(screen.queryByText(/invalid input/i)).not.toBeInTheDocument();
  });

  it('should match the snapshot without icon and error', () => {
    const { asFragment } = render(<TextField {...commonProps} />);

    expect(asFragment()).toMatchSnapshot();
  });

  it('should match the snapshot with an icon', () => {
    const { asFragment } = render(<TextField {...commonProps} icon={FaUser} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
