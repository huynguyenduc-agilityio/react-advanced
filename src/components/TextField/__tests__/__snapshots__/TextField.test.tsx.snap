// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TextField Component should match the snapshot with an icon 1`] = `
<DocumentFragment>
  <div
    class="chakra-form-control css-zol2i8"
    role="group"
  >
    <div
      class="css-h38yef"
    >
      <div
        class="css-jes1ca"
      >
        <svg
          class="chakra-icon css-vpsn6m"
          fill="currentColor"
          focusable="false"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 448 512"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M224 256c70.7 0 128-57.3 128-128S294.7 0 224 0 96 57.3 96 128s57.3 128 128 128zm89.6 32h-16.7c-22.2 10.2-46.9 16-72.9 16s-50.6-5.8-72.9-16h-16.7C60.2 288 0 348.2 0 422.4V464c0 26.5 21.5 48 48 48h352c26.5 0 48-21.5 48-48v-41.6c0-74.2-60.2-134.4-134.4-134.4z"
          />
        </svg>
        <label
          class="chakra-form__label css-1bat1t4"
          for="field-:r6:"
          id="field-:r6:-label"
        >
          Username
        </label>
      </div>
      <div
        class="css-1rr4qq7"
      >
        <input
          class="chakra-input css-0"
          id="field-:r6:"
          placeholder="Enter your username"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`TextField Component should match the snapshot without icon and error 1`] = `
<DocumentFragment>
  <div
    class="chakra-form-control css-zol2i8"
    role="group"
  >
    <div
      class="css-h38yef"
    >
      <div
        class="css-jes1ca"
      >
        <label
          class="chakra-form__label css-1bat1t4"
          for="field-:r5:"
          id="field-:r5:-label"
        >
          Username
        </label>
      </div>
      <div
        class="css-1rr4qq7"
      >
        <input
          class="chakra-input css-0"
          id="field-:r5:"
          placeholder="Enter your username"
        />
      </div>
    </div>
  </div>
</DocumentFragment>
`;
